allennlp/modules/transformer/t5.py

```diff
@@ -773,187 +801,473 @@ def forward(
 class T5EncoderStack(T5Stack, FromParams):
     def __init__(
         self,
-        layers: Optional[List[T5Block]] = None,
-        token_embeddings: Optional[nn.Embedding] = None,
+        token_embeddings: nn.Embedding,
+        blocks: List[T5Block],
         final_layer_norm: Optional[T5LayerNorm] = None,
         dropout: float = 0.1,
     ):
-        if layers is None:
-            layers = [
-                T5Block(
-                    attention=T5LayerSelfAttention(
-                        T5Attention(False, has_relative_attention_bias=(i == 0))
-                    ),
-                    cross_attention=None
-                )
-                for i in range(24)
-            ]
-        else:
-            if any(b.is_decoder for b in layers):
-                from allennlp.common.checks import ConfigurationError
-
-                raise ConfigurationError(
-                    "Found a decoder block in an encoder stack. This won't work."
-                )
+        if any(b.is_decoder for b in blocks):
+            raise ConfigurationError("Found a decoder block in an encoder stack. This won't work.")
+
+        super().__init__(
+            token_embeddings,
+            blocks,
+            final_layer_norm=final_layer_norm,
+            dropout=dropout,
+        )
 
-        super().__init__(layers, token_embeddings, final_layer_norm, dropout)
+    @classmethod
+    def basic_encoder(
+        cls,
+        token_embeddings: nn.Embedding,
+        num_blocks: int = 6,
+        block_self_attention: Lazy[T5Attention] = Lazy(T5Attention),
+        final_layer_norm: Optional[T5LayerNorm] = None,
+        block_ff: Lazy[T5LayerFF] = Lazy(T5LayerFF),
+        dropout: float = 0.1,
+    ) -> "T5EncoderStack":
+        blocks = [
+            T5Block(
+                attention=T5LayerSelfAttention(
+                    self_attention=block_self_attention.construct(
+                        is_decoder=False, has_relative_attention_bias=(i == 0)
+                    )
+                ),
+                cross_attention=None,
+                ff=block_ff.construct(),
+            )
+            for i in range(num_blocks)
+        ]
+        return cls(token_embeddings, blocks, final_layer_norm=final_layer_norm, dropout=dropout)
```

Comment:
This approach is my current best idea for how to support the different pre-trained dimensions without resorting to HF's config mechanism.